src/upgrade.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for handling Pgbouncer in-place upgrades."""
+
+import json
+import logging
+
+from charms.data_platform_libs.v0.upgrade import (
+    ClusterNotReadyError,
+    DataUpgrade,
+    DependencyModel,
+    KubernetesClientError,
+)
+from lightkube.core.client import Client
+from lightkube.core.exceptions import ApiError
+from lightkube.resources.apps_v1 import StatefulSet
+from ops.charm import WorkloadEvent
+from ops.pebble import ConnectionError
+from pydantic import BaseModel
+from typing_extensions import override
+
+DEFAULT_MESSAGE = "Pre-upgrade check failed and cannot safely upgrade"
+
+logger = logging.getLogger(__name__)
+
+
+class PgbouncerDependencyModel(BaseModel):
+    """Model for Pgbouncer Operator dependencies."""
+
+    charm: DependencyModel
+    rock: DependencyModel
+
+
+def get_pgbouncer_k8s_dependencies_model() -> PgbouncerDependencyModel:
+    """Return the Pgbouncer dependencies model."""
+    with open("src/dependency.json") as dependency_file:
+        _deps = json.load(dependency_file)
+    return PgbouncerDependencyModel(**_deps)
+
+
+class PgbouncerUpgrade(DataUpgrade):
+    """Implementation of :class:`DataUpgrade` overrides for in-place upgrades."""
+
+    def __init__(self, charm, model: BaseModel, **kwargs):
+        super().__init__(charm, model, **kwargs)
+        self.charm = charm
+
+        self.framework.observe(self.charm.on.upgrade_relation_changed, self._on_upgrade_changed)
+        self.framework.observe(
+            getattr(self.charm.on, "pgbouncer_pebble_ready"), self._on_pgbouncer_pebble_ready
+        )
+
+    def _cluster_checks(self) -> None:
+        """Check that the cluster is in healthy state."""
+        try:
+            if not self.charm.check_pgb_running():
+                raise ClusterNotReadyError(
+                    DEFAULT_MESSAGE, "Not all pgbouncer services are up yet."
+                )
+        except ConnectionError:
+            raise ClusterNotReadyError(DEFAULT_MESSAGE, "Not all pgbouncer services are missing.")
+
+        if self.charm.backend.postgres and not self.charm.backend.ready:
+            raise ClusterNotReadyError(DEFAULT_MESSAGE, "Backend relation is still initialising.")
+
+    @override
+    def pre_upgrade_check(self) -> None:
+        """Runs necessary checks validating the cluster is in a healthy state to upgrade.
+
+        Called by all units during :meth:`_on_pre_upgrade_check_action`.
+
+        Raises:
+            :class:`ClusterNotReadyError`: if cluster is not ready to upgrade.
+        """
+        self._cluster_checks()
+
+        try:
+            self._set_rolling_update_partition(self.charm.app.planned_units() - 1)
```

Comment:
Just a question: doesn't this trigger the upgrade in the last two units before it's needed to call the `resume-upgrade` action?

Reply:
It's a copy paste from the [Postgresql charm](https://github.com/canonical/postgresql-k8s-operator/blob/main/src/upgrade.py#L220). I haven't noticed it while testing and the integration test seems to upgrade the first unit only after `resume-upgrade` was issued.